File: server/app/helpers/api/directories/bad_entries.rb
class Server
  module App
    module Helpers

      def bad_entries( problem, dir_path, dir_config )

        dirname = File.basename dir_path

        entries = directory_entries dir_path, dir_config
        entries = entries.select do |entry|
          entry[:status] === problem
        end.map do |entry|
          entry[:name]
        end

        {
          type: :bad_entries,
          path: "#{ dir_path }/~dir",
          dirname: dirname,
          entries: entries,
          problem: problem,
        }

      end

    end
  end
end

File: client/app/views/bad_entries.js
app.views.bad_entries = ( r, data ) => (a,x) => {

  let entries = data.entries

  let message = data.problem === "missing" ? "Add missing entries?" : "Remove unknown entries?"
  let submitText = data.problem === "missing" ? "Create entries?" : "Delete entries?"

  return [
    a.h3( data.dirname ),
    message,
    a.ul(
      entries.map( function( entry ) {
        return a.li( entry )
      } )
    ),
    x.appkit.form( (f) => [
      f.button( {
        icon: "fa fa-times",
        text: "Cancel",
        buttonTag: { class: "btn btn-secondary" },
        onclick: (e, el, form) => r.open( `/${ data.path }` ),
      } ),
      " ",
      f.submit( {
        icon: "fa fa-check",
        text: submitText,
        buttonTag: { class: "btn btn-primary" },
      } ),
    ], {
      data: data,
      action: `/api/${ data.path }/${ data.problem }`,
      success: function( data, el ) {
        r.open( `/${ data.path }` )
      },
    } )
  ]


}
